### ISSUE TITLE
RFC: group commit

### ISSUE THREAD

**Issue by mqliang** [coreos/etcd#5724]
> The idea with group commit is to amortise the costs of each `fsync()` over multiple commits from multiple parallel transactions. If there are say 10 transactions in parallel trying to commit, we can force all of them to disk at once with a single `fsync()`, rather than do one `fsync()` for each. This can greatly reduce the need for `fsync()` calls, and consequently greatly improve the commits-per-second throughput.

Pros:
* less writing, more disk friendly
* throughput improving

Cons:
* larger write latency. But latency could be configurable, for example, write log every 10ms. 

@xiang90 @gyuho Does group commit worth doing?

**Comment by xiang90** [comment]
what is what we have already done. There are a few ways to improve our existing "batching" mechanism.  but i do not think it worth the effort right now.

**Comment by mqliang** [comment]
@xiang90 Mind to refer the code about the "batching" mechanism?

**Comment by mqliang** [comment]
@xiang90 May I ask the code snippet where we implement "group commit"? I look through the source code, we do have a [buffer](https://github.com/coreos/etcd/blob/master/wal/encoder.go#L33) in WAL encoder, but it seems used to reduce allocation when Marshal, instead of cache commit and sync to disk periodically.

**Comment by heyitsanthony** [comment]
@mqliang the batching happens at the raft level; multiple requests are batched into `[]raft.Ready.Entries` which commit in a single `fdatasync`.
